### ISSUE TITLE
`label:` does not support general variables

### ISSUE THREAD

**Issue by harelwa** [go-task/task#1023]
> Thanks for your bug report!
>
> Before submitting this issue, please make sure the same problem was
> not already reported by someone else.
>
> Please describe the bug you're facing. Consider pasting example
> Taskfiles showing how to reproduce the problem.

- Task version: v3.19.1 (h1:2KMJk6mDBacSPuFxPFvlvvHJwGZtU/hN2ENZpaFqR5s=)
- Operating System: MacOS (Arm64)

According to the [API ref](https://taskfile.dev/api/#task), `label` "Supports variables".

From what I've checked It looks like `label` supports only task variables passed by other tasks.

The following are not considered: 

- `Taskfile` level vars that requires evaluation and.or their default values
- `vars:` of the task itself
- env vars in generals are ignored ( just saying it would be useful to have them taken into account )

> I can work around env vars, but working around the rest is messy

`Taskfile.yml`:

```yaml
version: "3"
output: interleaved

vars:
  PYVERSION: "${PYVERSION}"

env:
  PYVERSION: "3.8"

tasks:
  default:
    cmds:
      - echo "PYVERSION ENV = ${PYVERSION}"
      - echo "PYVERSION VAR = {{.PYVERSION}}"

  foo:
    cmds:
      - echo "A" > env.yaml

  ya:
    deps:
      - foo
    desc: _
    cmds:
      - echo "{{.PYVERSION}}"
    label: 'ya-{{.PYVERSION}}'
    sources:
      - env.yaml
```

`commands`:

```bash
❯ task ya               
task: [foo] echo "A" > env.yaml
task: [ya-${PYVERSION}] echo "${PYVERSION}"
3.8
❯ task ya               
task: [foo] echo "A" > env.yaml
task: Task "ya-${PYVERSION}" is up to date
❯ task ya PYVERSION=3.10
task: [foo] echo "A" > env.yaml
task: [ya-3.10] echo "3.10"
3.10
❯ task ya PYVERSION=3.10
task: [foo] echo "A" > env.yaml
task: Task "ya-3.10" is up to date
❯ ls .task/checksum/
ya-3-10  ya---PYVERSION-
```

Many many thanks in advance,

Harel

**Comment by pd93** [comment]
The problem here is _**not**_ that `label` does not support variables. See the example below:

```yaml
vars:
  FOO: foo

tasks:
  default:
    label: "default-{{.FOO}}"
    cmds:
      - echo {{.FOO}}
```

```sh
❯ task
task: [default-foo] echo foo   # <--- This task has the expected label
foo
```

The problem is that `vars` will not interpret environment variables at runtime. In your example:

```yaml
vars:
  PYVERSION: "${PYVERSION}"
```

The variable `{{.PYVERISON}}` will just contain the raw string `${PYVERSION}` and not the actual value of the environment variable. Only templating variables can be used as labels, not environment variables.

There is an minor exception to this rule which can sometimes create confusion... Given the following Taskfile:

```yaml
env:
  BAR: bar

tasks:
  default:
    label: "default-{{.BAR}}"
    cmds:
      - echo {{.BAR}}
```

```sh
❯ task
task: [default-bar] echo bar # <--- This task has the expected label
bar
```

The label will be printed with the variable populated. However, this is only because variables specified with the `env` key are _also_ made available to the templating engine. This can be useful. However, it comes with the caveat that you cannot override the value at runtime:

```sh
❯ BAR=overridden task        # <--- Attempt to override variable
task: [default-bar] echo bar # <--- The task label is not overridden
bar
```

It is also worth noting that you cannot directly use the environment variables in the label either. In the example below, I'm using `$BAR` instead of `{{.BAR}}`:

```yaml
env:
  BAR: bar

tasks:
  default:
    label: "default-$BAR"
    cmds:
      - echo $BAR
```

```sh
❯ task
task: [default-$BAR] echo $BAR # <--- label is not interpreted
bar                            # <--- value is correctly printed when the command is run
```

And if we attempt to override...

```sh
❯ BAR=overridden task
task: [default-$BAR] echo $BAR # <--- label is still not interpreted
overridden                     # <--- value is correctly printed when the command is run
```

We have the same issue.

Unfortunately, changes to how variables work are unlikely to happen in the current major version of Task (due to backward incompatibilities). However, @andreynering might have more to say on this than me.

## TL;DR

Labels _do_ support variables. However, there is no way to use environment variables in your label (directly or indirectly via templating variables). You can only directly use templating variables.

**Comment by harelwa** [comment]
thank you @pd93 for this elaborate answer. 

i''l elaborate a bit more about my use case, so you have it mind - 

- my use case of `label` is for caching / checksum ( not printing ).
- eventually, i need to somehow insert an `env var` value to the label, without calling the task with an input `var`
- i need this `env var` to have a *default value*, which is globally available in `Taskfile.yml`
 
here is one way which does not work, and perhaps you could tell me why -

`Taskfile.yml`:

```yaml
version: "3"
output: interleaved

env:
  PYVERSION: 3.7 # default value. must be here, as it's used "globally"

tasks:
  foo:
    cmds:
      - echo "A" > env.yaml

  ya:
    desc: _
    deps:
      - foo
    vars:
      _PYVERSION:
        sh: echo ${PYVERSION}
    cmds:
      - echo "PYVERSION  ENV = ${PYVERSION}"
      - echo "_PYVERSION VAR = {{._PYVERSION}}"
    label: 'ya-{{._PYVERSION}}'
    sources:
      - env.yaml
```

Usage:

```bash
❯ task ya
task: [foo] echo "A" > env.yaml
task: [ya-] echo "PYVERSION  ENV = ${PYVERSION}"
PYVERSION  ENV = 3.7
task: [ya-] echo "_PYVERSION VAR = "
_PYVERSION VAR =  # this is what i don't understand

❯ PYVERSION=3.10 task ya
task: [foo] echo "A" > env.yaml
task: [ya-3.10] echo "PYVERSION  ENV = ${PYVERSION}"
PYVERSION  ENV = 3.10
task: [ya-3.10] echo "_PYVERSION VAR = 3.10"
_PYVERSION VAR = 3.10

❯ PYVERSION=3.7 task ya  # this is problematic for what i'm trying to achieve, as it's suppose to not run to use the already cached default value
task: [foo] echo "A" > env.yaml
task: [ya-3.7] echo "PYVERSION  ENV = ${PYVERSION}"
PYVERSION  ENV = 3.7
task: [ya-3.7] echo "_PYVERSION VAR = 3.7"
_PYVERSION VAR = 3.7
```

regardless, here is `env` / `var` the confusion "summary" :

`Taskfile.yml`:

```yaml
version: "3"
output: interleaved

vars:
  PYVERSION: "${PYVERSION}"

env:
  PYVERSION: "3.8"

tasks:
  show-confusion:
    desc: _
    label: "show-confusion-{{.PYVERSION}}"
    cmds:
      - echo "PYVERSION ENV = ${PYVERSION}"
      - echo "PYVERSION VAR = {{.PYVERSION}}"
```

`Usage:`

```bash
❯ PYVERSION=4.8 task show-confusion
task: [show-confusion-${PYVERSION}] echo "PYVERSION ENV = ${PYVERSION}"
PYVERSION ENV = 4.8
task: [show-confusion-${PYVERSION}] echo "PYVERSION VAR = ${PYVERSION}"
PYVERSION VAR = 4.8
```

Thanks,

Harel

**Comment by harelwa** [comment]
I'll use `variables` for now, no env vars. thank you
